pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -202,10 +239,53 @@ func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, pus
 		model.XdsLogDetails{AdditionalInfo: fmt.Sprintf("cached:%v/%v", clusterCacheStats.hits, clusterCacheStats.hits+clusterCacheStats.miss)}, true
 }
 
+// getRemovedClusterName finds the name of the removed cluster given a hostname and list of cluster names.
+// subsets is ignored if it is nil
+func getRemovedClusterNames(resNames []string, subsets []string, target string) []string {
+	removed := make([]string, 0)
+	for _, n := range resNames {
+		_, sub, svcHost, _ := model.ParseSubsetKey(n)
+		if svcHost == host.Name(target) && (len(subsets) == 0 || contains(subsets, sub)) {
+			removed = append(removed, n)
+		}
+	}
+	return removed
+}
+
+// getDeletedSubsets detects deleted subsets and returns those names
+func getDeletedSubsets(orig []*networking.Subset, new []*networking.Subset) []string {
+	subsetNames := make([]string, 0)
+	// deleted if exists in orig but not in new
+	ne := make(map[string]struct{})
```

Comment:
use sets.Set?

Reply:
maybe, but then I would have to extract everything into []string...is it ok to keep this way?

Reply:
https://github.com/istio/istio/blob/master/pilot/pkg/util/sets/string.go#L19 - it is same as what you are doing here. But a cleaner api